web/html/src/manager/virtualization/guests/console/guests-console.js

```diff
@@ -132,34 +170,74 @@ class GuestsConsole extends React.Component<Props, State> {
     this.setState({ password: model.password });
   }
 
-  render() {
-    const buttonValues = {
-      askPassword: {
-        text: t('Submit'),
+  refreshToken = () => {
+    Network.post(`/rhn/manager/api/systems/details/virtualization/guests/consoleToken/${this.props.guestUuid}`,
+                 this.state.currentToken, "application/json").promise.then(
+      response => {
+        this.popupSubmit = undefined;
+        this.setState({
+          connected: false,
+          currentToken: response,
+          popupState: 'wait',
+        }, this.connect);
       },
-      errors: {
-        text: t('Retry'),
-      },
-    }[this.state.popupState];
-    const canResize = this.client != null && this.client.canResize;
-    const popupContent = () => {
-      if (this.state.popupState === 'wait') {
-        return <Loading text={t('Connecting...')} withBorders={false} />;
-      }
-      if (this.state.popupState === 'askPassword') {
-        return (
-          <Form model={this.state} className="form-horizontal" onChange={this.onPasswordChange}>
-            <Password
-              name="password"
-              label={t('Password')}
-              labelClass="col-md-3"
-              divClass="col-md-6"
-            />
-          </Form>
-        );
+      xhr => {
+        if (xhr.status === 400) {
+          // We have hit a time where the VM is not yet in the DB, retry
+          this.refreshToken();
+        } else {
+          this.popupSubmit = () => {
+            window.location.reload();
+          }
+          this.setState({
+            error: Network.errorMessageByStatus(xhr.status),
+            popupState: 'errors',
+          }, this.showPopup)
+        }
       }
-      return this.state.error;
-    };
+    );
+  };
+
+  virtEventHandler = (msg: Object) => {
+    if (msg.event === 'resumed' && msg.detail === 'migrated') {
+      this.client = undefined;
+      setTimeout(() => {}, 500);
+      this.refreshToken();
+      this.setState({vmState: 'running', hostId: msg.hostId}, this.showPopup);
+    }
+
+    // Ignore all other events that come from other hosts
+    if (this.state.hostId !== msg.hostId) {
+      return;
+    }
+
+    if (msg.event === 'started') {
+      setTimeout(() => {}, 500);
+      this.refreshToken();
+      this.setState({vmState: 'running'}, this.showPopup);
+      return;
+    }
+
+    if (msg.event === 'suspended' && msg.detail === 'migrated') {
+      this.client.removeErrorHandler();
+      this.setState({vmState: 'migrating'});
+      return;
+    }
+
+    if (msg.event === 'shutdown') {
+      this.client.removeErrorHandler();
+      return;
+    }
+
+    if (msg.event === 'stopped') {
+      this.client = undefined;
+      this.setState({vmState: 'stopped'});
+      return;
+    }
```

Comment:
I am personally not a fun of detached conditions and the `return` behavior cutting the statement in the middle. I prefer the reading logic of `if....else if.....else if....else` then I know what's the reasoning, and eventually the `return` at the end. But it is just my personal opinion in this, I don't see any drawback in both cases.

Reply:
hehe, it's a matter of taste indeed :)